dcm2niix_gear/dcm2niix/arrange.py

```diff
@@ -118,6 +118,118 @@ def exit_if_archive_empty(archive_obj):
+    file_name_path_dict = {}
+    file_tree = []
+    gen = os.walk(dir_loc, topdown=True)
+    next_gen = next(gen)
+    # want to exclude files matching certain patterns str_to_exclude
+    strs_to_exclude = [".DS_Store", "^._"]
+    excludes = [re.compile(s) for s in strs_to_exclude]
+    while next_gen != StopIteration:
+
+        # info on current folder and contents
+        (loc_i, dirs_i, files_i) = next_gen
+
+        for file_ij in files_i:
+
+            if all([not (exclude.search(file_ij)) for exclude in excludes]):
+
+                # avoid collisions from collapsing filepaths
+                if file_ij in file_name_path_dict.keys():
+                    print(f"more than one file with name of {file_ij} in directory tree, exiting.")
+                    os.sys.exit(1)
+
+                # set of files (short paths or leaves) found in directory
+                file_set = file_set | set([file_ij])
+
+                # dictionary to look up full file paths
+                file_name_path_dict[file_ij] = os.path.join(loc_i, file_ij)
+
+            # directory structure in list form, no filter on files
+            file_tree.append(loc_i + "/" + file_ij)
+
+        # display
+        if print_out:
+            print(f"file_set: {file_set}")
+            print(f"file_tree: {file_tree}")
+            print(f"file_name_path_dict: {file_name_path_dict}")
+
+        # next folder:
+        try:
+            next_gen = next(gen)
+        except StopIteration:
+            next_gen = StopIteration
+
+    return \
+        file_set, \
+        file_tree, \
+        file_name_path_dict
+
+
+def flatten_directory(dir_source, dir_target, overwrite=False):
+    """
+    Args:
+        dir_source: possibly nested source
+        dir_target: place to create flat file structure
+
+    Returns:
+        info on source directory:
+            file_set_source
+            file_tree_source
+            file_name_path_dict_source
+
+        info on target directory
+            file_set_target
+            file_tree_target
+            file_name_path_dict_target
+    """
+    # TODO maybe write test for overwriting
+
+    # source directory tally
+    file_set_source, file_tree_source, file_name_path_dict_source = \
+        tally_files(dir_source)
+
+    # check for already existing/clear way for target directory
+    if os.path.exists(dir_target):
+        if overwrite:
+            shutil.rmtree(dir_target)
+        else:
+            print(f"file {dir_target} already exists, exiting.")
+            os.sys.exit(1)
+    os.mkdir(dir_target)
+
+    # the main part, move the files from the possibly hierarchical structure
+    # to the flat structure
+    for file in file_name_path_dict_source.keys():
+        shutil.move(file_name_path_dict_source[file], dir_target)
+
+    # target directory tally
+    file_set_target, file_tree_target, file_name_path_dict_target = \
+        tally_files(dir_target)
+
+    return \
+        file_set_source, \
```

Comment:
Just wondering why we need source directory info here, since all the files have been moved to the target?

Reply:
I think just for QA purposes, open to other suggestions :)